### ISSUE TITLE
Shouldn't this be synchronous replication

### ISSUE THREAD

**Issue by danielwestermann** [citusdata/pg_auto_failover#87]
Hi,

 I just followed the README to setup the three nodes with PostgreSQL 12:
```
postgres@pg-af1:/u02/pgdata/12/PG1/ [PG1] pg_autoctl show state
Name |   Port | Group |  Node |     Current State |    Assigned State
---------------------------+--------+-------+-------+-------------------+------------------
pg-af1.it.dbi-services.com |   5432 |     0 |     1 |      wait_primary |      wait_primary
pg-af2.it.dbi-services.com |   5432 |     0 |     2 |        catchingup |        catchingup
```

What does the "catchingup" state really mean? I'd expected something like "in sync". When checking synchronous_standby_names I'd expected to see synchronous replication which apparently is not the case:

```
postgres@pg-af1:/u02/pgdata/12/PG1/ [PG1] sq
psql (12.0 dbi services build)
Type "help" for help.

postgres=# show synchronous_standby_names ;
synchronous_standby_names
---------------------------

(1 row)
```

Am I missing something? Replication in general is fine:

```
postgres=# select * from pg_stat_replication ;
-[ RECORD 1 ]----+------------------------------
pid              | 8336
usesysid         | 16385
usename          | pgautofailover_replicator
application_name | walreceiver
client_addr      | 192.168.22.71
client_hostname  | pg-af2.it.dbi-services.com
client_port      | 48680
backend_start    | 2019-10-19 15:39:00.508099+02
backend_xmin     |
state            | streaming
sent_lsn         | 0/30017A0
write_lsn        | 0/30017A0
flush_lsn        | 0/30017A0
replay_lsn       | 0/30017A0
write_lag        |
flush_lag        |
replay_lag       |
sync_priority    | 0
sync_state       | async
reply_time       | 2019-10-19 15:49:16.594248+02
```

Thanks
Daniel

**Comment by DimCitus** [comment]
Hi Daniel! Thanks for your interest in pg_auto_failover and for your patience while the team was busy and traveling!

The CATCHINGUP state is documented with the others at https://pg-auto-failover.readthedocs.io/en/latest/fsm.html#state-reference and we can read:

> The monitor assigns catchingup to the standby node when the primary is ready for a replication connection (pg_hba.conf has been properly edited, connection role added, etc).
> 
> The standby node keeper runs pg_basebackup, connecting to the primary’s nodename and port. The keeper then edits recovery.conf and starts PostgreSQL in hot standby node.

So the replication is now ongoing and the monitor needs some more information before it can ask the primary to add the secondary to the `synchronous_standby_names`. Namely:
  1. the monitor health check must succeed, you can have a look at `pgautofailover.node` table for more information
  2. the WAL lag must be kept under 1 WAL segment, by default, you can change this setting on the Monitor side (https://pg-auto-failover.readthedocs.io/en/latest/ref/configuration.html#pg-auto-failover-monitor).

I guess that one of those elements has not converged to the acceptable state yet at the moment when you're still *catchingup*. As soon as those will resolve, the monitor will assign roles *primary* and *secondary* and the replication will switch to SYNC.

**Comment by danielwestermann** [comment]
Hi Dimitri,

when installing from source and you are on CentOS/RedHat 8 you should add the following to the documentation, otherwise the systemd service will not start:

```
sudo semanage fcontext -a -t bin_t [PATH_TO]/pg_autoctl
restorecon -v [PATH_TO]/pg_autoctl
```

I am still struggling with the state:

```
postgres@pg-af1:/home/postgres/ [PG1] pg_autoctl show state
Name |   Port | Group |  Node |     Current State |    Assigned State
---------------------------+--------+-------+-------+-------------------+------------------
pg-af1.it.dbi-services.com |   5432 |     0 |     1 |      wait_primary |      wait_primary
pg-af2.it.dbi-services.com |   5432 |     0 |     2 |        catchingup |        catchingup
```

This is the content of the node table:
```
pg_auto_failover=# \x
Expanded display is on.
pg_auto_failover=# select * from pgautofailover.node;
-[ RECORD 1 ]-------+------------------------------
formationid         | default
nodeid              | 2
groupid             | 0
nodename            | pg-af2.it.dbi-services.com
nodeport            | 5432
goalstate           | catchingup
reportedstate       | catchingup
reportedpgisrunning | t
reportedrepstate    | unknown
reporttime          | 2019-10-30 20:51:45.858825+01
reportedlsn         | 0/40005D8
walreporttime       | 2019-10-30 20:51:45.858825+01
health              | 0
healthchecktime     | 2019-10-30 20:51:35.752691+01
statechangetime     | 2019-10-30 20:51:45.858825+01
-[ RECORD 2 ]-------+------------------------------
formationid         | default
nodeid              | 1
groupid             | 0
nodename            | pg-af1.it.dbi-services.com
nodeport            | 5432
goalstate           | wait_primary
reportedstate       | wait_primary
reportedpgisrunning | t
reportedrepstate    | async
reporttime          | 2019-10-30 20:51:43.847564+01
reportedlsn         | 0/40005D8
walreporttime       | 2019-10-30 20:51:43.847564+01
health              | 1
healthchecktime     | 2019-10-30 20:51:29.776176+01
statechangetime     | 2019-10-30 20:51:43.847564+01

pg_auto_failover=#
```

... which does not give me any hint what might be wrong. From a pure PostgreSQL perspective replication is up to date:

```
postgres=# select * from pg_stat_replication ;
-[ RECORD 1 ]----+------------------------------
pid              | 1262
usesysid         | 16385
usename          | pgautofailover_replicator
application_name | walreceiver
client_addr      | 192.168.22.71
client_hostname  | pg-af2.it.dbi-services.com
client_port      | 60656
backend_start    | 2019-10-30 20:45:21.912271+01
backend_xmin     |
state            | streaming
sent_lsn         | 0/40005D8
write_lsn        | 0/40005D8
flush_lsn        | 0/40005D8
replay_lsn       | 0/40005D8
write_lag        |
flush_lag        |
replay_lag       |
sync_priority    | 0
sync_state       | async
reply_time       | 2019-10-30 20:54:07.05031+01
```

No lag, all lsn columns up to date. I am not sure what prevents the state from changing. 

Best regards
Daniel

**Comment by dimitri** [comment]
I'm not sure where to add the extra SElinux setup commands in the documentation. Would that be somewhere in https://pg-auto-failover.readthedocs.io/en/latest/install.html#installing-a-pgautofailover-systemd-unit then?

**Comment by danielwestermann** [comment]
Hi Dimitri,

yes, I think the "Installing a pgautofailover Systemd unit" would be the right place to add this, maybe as a note.

Best regards
Daniel



________________________________

**Comment by danielwestermann** [comment]
Now I remember why I used the --nodename parameter:


```
postgres@pg-af3 ~]$ ./_reset.sh 
17:51:00 INFO  Using --nodename "pg-af3", which resolves to IP address "192.168.22.72"
17:51:02 INFO  Initialising a PostgreSQL cluster at "/u02/pgdata/12/af"
17:51:02 INFO   /u01/app/postgres/product/12/db_0/bin/pg_ctl --pgdata /u02/pgdata/12/af --options "-p 5432" --options "-h *" --wait start
17:51:03 INFO  Granting connection privileges on 192.168.22.0/24
17:51:03 INFO  Your pg_auto_failover monitor instance is now ready on port 5432.
17:51:03 INFO  pg_auto_failover monitor is ready at postgres://autoctl_node@pg-af3:5432/pg_auto_failover
17:51:03 INFO  Monitor has been succesfully initialized.
17:51:03 INFO  Found pg_ctl for PostgreSQL 12.0 at /u01/app/postgres/product/12/db_0/bin/pg_ctl
17:51:03 INFO  Registered node pg-af1.it.dbi-services.com:5432 with id 1 in formation "default", group 0.
17:51:03 INFO  Writing keeper init state file at "/home/postgres/.local/share/pg_autoctl/u02/pgdata/12/PG1/pg_autoctl.init"
17:51:03 INFO  Successfully registered as "single" to the monitor.
17:51:05 INFO  Initialising a PostgreSQL cluster at "/u02/pgdata/12/PG1"
17:51:05 INFO  Postgres is not running, starting postgres
17:51:05 INFO   /u01/app/postgres/product/12/db_0/bin/pg_ctl --pgdata /u02/pgdata/12/PG1 --options "-p 5432" --options "-h *" --wait start
17:51:05 INFO  CREATE DATABASE postgres;
17:51:05 INFO  The database "postgres" already exists, skipping.
17:51:06 INFO  FSM transition from "init" to "single": Start as a single node
17:51:06 INFO  Initialising postgres as a primary
17:51:06 INFO  Transition complete: current state is now "single"
17:51:06 INFO  Keeper has been succesfully initialized.
17:51:06 WARN  Failed to connect: Permission denied
17:51:06 FATAL Failed to find a local IP address, please provide --nodename.
17:51:06 FATAL Failed to auto-detect the hostname of this machine, please provide one via --nodename
17:51:07 WARN  Failed to connect: Permission denied
17:51:07 FATAL Failed to find a local IP address, please provide --nodename.
17:51:07 FATAL Failed to auto-detect the hostname of this machine, please provide one via --nodename
                      Name |   Port | Group |  Node |     Current State |    Assigned State
---------------------------+--------+-------+-------+-------------------+------------------
pg-af1.it.dbi-services.com |   5432 |     0 |     1 |            single |            single

[postgres@pg-af3 ~]$ 
```

Without the nodename the initialization of the replica fails with:
```

17:51:06 WARN  Failed to connect: Permission denied
17:51:06 FATAL Failed to find a local IP address, please provide --nodename.
17:51:06 FATAL Failed to auto-detect the hostname of this machine, please provide one via --nodename
```

**Comment by DimCitus** [comment]
Closing this issue for triaging purposes. Feel free to re-open of more actions are expected.
